exampleservice/CustomerService/Handler/RegisterCustomerHandler.cs

```diff
@@ -0,0 +1,78 @@
+﻿using exampleservice.CustomerService.Contract;
+using exampleservice.CustomerService.Controller;
+using exampleservice.CustomerService.Events;
+using exampleservice.CustomerService.Utils;
+using exampleservice.Framework.Abstract;
+using exampleservice.Framework.BaseFramework;
+using System;
+using System.Threading.Tasks;
+
+namespace exampleservice.CustomerService.Handler
+{
+    internal class RegisterCustomerHandler : CustomerHandlerBase<RegisterCustomerCommand>
+    {
+        private IMessageBus bus;
+        private ICustomerServiceDataBaseRepository dataBaseRepository;
+
+        public RegisterCustomerHandler(IMessageBus bus, ICustomerServiceDataBaseRepository dataBaseRepository)
+        {
+            this.bus = bus ?? throw new ArgumentNullException(nameof(bus));
+            this.dataBaseRepository = dataBaseRepository ?? throw new ArgumentNullException(nameof(dataBaseRepository));
+        }
+
+        internal override async Task<EventBase> Handle(RegisterCustomerCommand command)
+        {
+            this.VerifyIputArguments(command);
+
+            string passwordHash = Password.ComputeHash(command.Customer.Password);
+            command.Customer.Password = null; //set PW to null asap to avoid saving or returning cleartext password at all cost
+
+            //check if customer username is already taken
+            CustomerSpecification customer = await dataBaseRepository.LoadCustomer(command.Customer.Username);
+            if (customer != null)
+            {
+                return new CustomerRegistrationFailedEvent() { Customer = command.Customer };
+            }
+            
+            command.Customer.PasswordHash = passwordHash;
+            command.Customer.CustomerId = Guid.NewGuid();
+            
+            if (await dataBaseRepository.CreateCustomer(command.Customer) > 0)
+            {
+                CustomerRegisteredEvent e = new CustomerRegisteredEvent() { Customer = command.Customer };
+                bus.PublishEvent(e);
```

Comment:
was this intentional? This can lead to funny effects because the publishing would occur after the return, all potential exceptions would be lost (afaik).